sql/src/main/java/io/crate/analyze/expressions/ExpressionToColumnIdentVisitor.java

```diff
@@ -0,0 +1,99 @@
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze.expressions;
+
+import io.crate.metadata.ColumnIdent;
+import io.crate.sql.tree.AstVisitor;
+import io.crate.sql.tree.Node;
+import io.crate.sql.tree.QualifiedNameReference;
+import io.crate.sql.tree.StringLiteral;
+import io.crate.sql.tree.SubscriptExpression;
+import org.elasticsearch.common.inject.internal.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+public class ExpressionToColumnIdentVisitor extends AstVisitor<ColumnIdent, List<String>> {
+
+    private final static ExpressionToColumnIdentVisitor INSTANCE = new ExpressionToColumnIdentVisitor();
+
+    private ExpressionToColumnIdentVisitor() {
+    }
+
+    public static ColumnIdent convert(Node node) {
+        return INSTANCE.process(node, null);
+    }
+
+    @Override
+    protected ColumnIdent visitQualifiedNameReference(QualifiedNameReference node, @Nullable List<String> context) {
+        if (node.getName().getParts().size() > 1) {
+            throw new IllegalArgumentException(String.format(
+                Locale.ENGLISH,
+                "Qualified name references are not allowed to contain paths ('%s')",
+                node.getName().toString()
+            ));
+        }
+        if (context != null) {
+            return new ColumnIdent(node.getName().toString(), context);
+        }
+        return new ColumnIdent(node.getName().toString());
+    }
+
+    @Override
+    protected ColumnIdent visitStringLiteral(StringLiteral node, List<String> context) {
+        if (context == null) {
+            // TopLevel doesn't allow string literals.
+            throw new IllegalArgumentException(String.format(
+                Locale.ENGLISH,
+                "String literal '%s' is not allowed as column identifier",
+                node.getValue()
+            ));
+        }
+
+        context.add(node.getValue());
+        return null;
+    }
+
+    @Override
+    protected ColumnIdent visitSubscriptExpression(SubscriptExpression node, List<String> context) {
+        if (node.index() instanceof QualifiedNameReference) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ENGLISH, "Index of subscript must not be a reference")
```

Comment:
caling this `Index` sounds a bit odd to me as this we use the term `index` only on array subscript. I think calling id `key` is better but not perfect, could still be meant as an array subscript ;)